### PR TITLE
[concept] add livez/readyz for etcd

### PR DESCRIPTION
**Pull request by logicalhan**
This is a prototype for adding livez/readyz support to etcd. Currently I've configured the `NO_SPACE` alarm to only count towards `\readyz` since it means etcd is degraded. Whether quorum should be included in liveness is an open question.

### CODE REVIEW

File: server/etcdserver/api/etcdhttp/health.go [hunk 1]
```diff
@@ -46,33 +49,59 @@ type ServerHealth interface {
...
+
+// HandleLivez registers metrics and health handlers. it checks health by using v3 range request
+// and its corresponding timeout.
+func HandleLivez(lg *zap.Logger, mux *http.ServeMux, srv ServerHealth) {
+	mux.Handle(PathLivez, NewHealthHandler(lg, func(excludedAlarms AlarmSet, serializable bool, endpoint string) Health {
+		if h := checkAlarms(lg, srv, excludedAlarms, endpoint); h.Health != "true" {
+			return h
+		}
+		// TODO(logicalhan) should we require quorum for livez?
+		return checkAPI(lg, srv, serializable)
```
Comment (ahrtr) [inline server/etcdserver/api/etcdhttp/health.go:71]:
```suggestion
		return checkAPI(lg, srv, true)
```

File: server/etcdserver/api/etcdhttp/health.go [hunk 2]
```diff
@@ -46,33 +49,59 @@ type ServerHealth interface {
...
+}
+
+// HandleReadyz registers metrics and health handlers. it checks health by using v3 range request
+// and its corresponding timeout.
+func HandleReadyz(lg *zap.Logger, mux *http.ServeMux, srv ServerHealth) {
+	mux.Handle(PathReadyz, NewHealthHandler(lg, func(excludedAlarms AlarmSet, serializable bool, endpoint string) Health {
+		if h := checkAlarms(lg, srv, excludedAlarms, endpoint); h.Health != "true" {
+			return h
+		}
+		return checkAPI(lg, srv, serializable)
```
Comment (ahrtr) [inline server/etcdserver/api/etcdhttp/health.go:82]:
```suggestion
		return checkAPI(lg, srv, false)
```

File: server/etcdserver/api/etcdhttp/health.go (hunk 1 above)
Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/health.go:71]: `livez` doesn't require quorum

Comment (logicalhan) [reply server/etcdserver/api/etcdhttp/health.go:71]: You mean it shouldn't?

Comment (logicalhan) [reply server/etcdserver/api/etcdhttp/health.go:71]: I'm confused, do you mean 'false' here and 'true' on line 82?

Comment (logicalhan) [reply server/etcdserver/api/etcdhttp/health.go:71]: Also I have a todo here for a specific reason. I'm not even sure that this code is doing the right thing, since it's going through the etcdclient and therefore can hit any particular etcd member. So even if serializable is set to false and we're getting stale reads from a particular etcd member, we're not actually guaranteed to hit our own etcd member, which is the thing we need to be testing for.

Comment (ahrtr) [reply server/etcdserver/api/etcdhttp/health.go:71]: Right. `livez` just check local instance's health. We shouldn't restart the local instance if the cluster is NOT health, e.g. quorum isn't satisfied.

Comment (logicalhan) [reply server/etcdserver/api/etcdhttp/health.go:71]: Yes, but we shouldn't even be going through the etcd client, is my point, since it load balances.

Comment (logicalhan) [reply server/etcdserver/api/etcdhttp/health.go:71]: Ah, it's not going through the client if serializable is false. I see.

**Review by serathius — CHANGES_REQUESTED** [review]
Don't want to rash into adding livez/readyz probe. Main problem with existing health probe we just added it to have it without proper consideration. 

I want livez to properly reflect fact that etcd needs restart, for example etcd is stuck on stalled storage https://docs.google.com/document/d/1U9hAcZQp3Y36q_JFiw2VBJXVAo2dK2a-8Rsbqv3GgDo/edit?usp=sharing. 

Readyz should properly reflect fact that etcd is ready to serve traffic. Don't think alarms matter here. It's a degradation, however it doesn't mean we shouldn't serve reads.

TLDR; I would like to have a design written that will do a proper analysis etcd failure modes and propose matching probes to detect them. Example https://github.com/kubernetes-sigs/metrics-server/issues/542

File: server/etcdserver/api/etcdhttp/health.go [hunk 3]
```diff
@@ -141,7 +169,7 @@ func getSerializableFlag(r *http.Request) bool {
 
 // TODO: etcdserver.ErrNoLeader in health API
 
-func checkAlarms(lg *zap.Logger, srv ServerHealth, excludedAlarms AlarmSet) Health {
+func checkAlarms(lg *zap.Logger, srv ServerHealth, excludedAlarms AlarmSet, healthType string) Health {
```
Comment (ahrtr) [inline server/etcdserver/api/etcdhttp/health.go:172]: The `healthType string` isn't used at all, can we remove it? 
```suggestion
func checkAlarms(lg *zap.Logger, srv ServerHealth, excludedAlarms AlarmSet) Health {
```

**Comment by jmhbnz** [comment]
Discussed during sig-etcd triage meeting. This original concept has now been superceeded by the work from @siyuanfoundation 🎉

**Comment by siyuanfoundation** [comment]
the work is tracked in https://github.com/etcd-io/etcd/issues/16007
